File: DraftRightMobile/android/app/src/main/kotlin/com/draftright/keyboard/KeyFeedback.kt
package com.draftright.keyboard

import android.content.Context
import android.media.AudioManager
import android.view.HapticFeedbackConstants
import android.view.View

/** The feedback-relevant class of a key press — decides which click plays. */
enum class KeyFeedbackKind { CHAR, SPACE, DELETE, ENTER, OTHER }

/**
 * Samsung-parity key feedback: a short haptic tick + a key-appropriate click on
 * every key actuation. Both effects go through the platform APIs that already
 * respect the user's OS settings — [View.performHapticFeedback] honours the
 * system haptic toggle, and [AudioManager.playSoundEffect] is silent when
 * "touch sounds" is off — so there is no separate settings gate to maintain.
 *
 * This is the single chokepoint every key routes through (Rule #1: feedback is
 * cross-cutting, so it lives in one place invoked from the touch handler, not
 * copied per key). The kind→sound mapping is pure and unit-tested
 * ([KeyFeedbackTest]); the firing itself is verified on-device.
 *
 * @param host any attached keyboard view — used as the haptic source; a single
 *             host works for every key since haptics aren't spatial.
 */
class KeyFeedback(private val host: View) {

    // Nullable: a handful of OEM/emulator builds expose no AUDIO_SERVICE. Missing
    // audio must degrade to "haptic only", never crash the keystroke.
    private val audio: AudioManager? =
        host.context.getSystemService(Context.AUDIO_SERVICE) as? AudioManager

    /** Fire haptic + sound for the key with the given [code]. Safe to call on every press. */
    fun onKey(code: Int) {
        // KEYBOARD_TAP is the constant OS keyboards use; it obeys the system
        // haptic setting without us reading it.
        host.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP)
        audio?.playSoundEffect(soundEffect(kindOf(code)))
    }

    companion object {
        /** Classify a key [code] (as used by [SpecialKeys]) into a feedback kind. */
        fun kindOf(code: Int): KeyFeedbackKind = when {
            SpecialKeys.isSpace(code) -> KeyFeedbackKind.SPACE
            code == SpecialKeys.BACKSPACE -> KeyFeedbackKind.DELETE
            code == SpecialKeys.ENTER -> KeyFeedbackKind.ENTER
            SpecialKeys.isCharKey(code) -> KeyFeedbackKind.CHAR
            else -> KeyFeedbackKind.OTHER // shift / layer switch / globe
        }

        /**
         * Platform sound-effect id for a kind. Matches Samsung/Gboard: letters
         * click "standard", space / delete / return have their own tones.
         */
        fun soundEffect(kind: KeyFeedbackKind): Int = when (kind) {
            KeyFeedbackKind.SPACE -> AudioManager.FX_KEYPRESS_SPACEBAR
            KeyFeedbackKind.DELETE -> AudioManager.FX_KEYPRESS_DELETE
            KeyFeedbackKind.ENTER -> AudioManager.FX_KEYPRESS_RETURN
            KeyFeedbackKind.CHAR, KeyFeedbackKind.OTHER -> AudioManager.FX_KEYPRESS_STANDARD
        }
    }
}

File: DraftRightMobile/ios/DraftRightKeyboard/KeyFeedback.swift
import UIKit
import DraftRightKeyboardCore

/// Samsung-parity key feedback for the iOS keyboard (#209): a system input click
/// on every key, plus a short haptic tick whose strength depends on the key kind
/// (see `KeyFeedbackKind.impact`). The single chokepoint every key routes through
/// — RULE #1: feedback is cross-cutting, so it lives here and is called from each
/// delegate method, not copied per key.
///
/// - The click uses `UIDevice.playInputClick()`, which is silent unless the user
///   enabled keyboard clicks and the input view conforms to
///   `UIInputViewAudioFeedback` (see `KeyboardViewController`). No settings gate
///   to maintain.
/// - Haptics require **Full Access** in a keyboard extension; without it we
///   degrade to click-only rather than doing nothing or crashing. The kind→sound
///   mapping is unit-tested (`KeyFeedbackKindTests`); the firing is verified
///   on-device.
struct KeyFeedback {
    /// Whether the extension has Full Access (from `UIInputViewController.hasFullAccess`).
    /// Read at fire time so toggling access mid-session is honoured.
    let hasFullAccess: () -> Bool

    /// Reused generators so each press doesn't allocate; `prepare()` on touch keeps
    /// the Taptic Engine warm for a crisp tick.
    private let lightHaptic = UIImpactFeedbackGenerator(style: .light)
    private let mediumHaptic = UIImpactFeedbackGenerator(style: .medium)

    init(hasFullAccess: @escaping () -> Bool) {
        self.hasFullAccess = hasFullAccess
    }

    /// Fire click + haptic for a key of the given [kind]. Safe on every press.
    func fire(_ kind: KeyFeedbackKind) {
        UIDevice.current.playInputClick()
        guard hasFullAccess() else { return }
        switch kind.impact {
        case .light: lightHaptic.impactOccurred()
        case .medium: mediumHaptic.impactOccurred()
        }
    }
}
